src/main/kotlin/com/shortener/data/cache/UrlEntryRepositoryCacheImpl.kt

```diff
@@ -0,0 +1,45 @@
+package com.shortener.data.cache
+
+import com.shortener.data.domain.EncodedSequence
+import com.shortener.data.domain.UrlEntry
+import com.shortener.data.repository.UrlEntryRepositoryBase
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.cache.annotation.CacheEvict
+import org.springframework.cache.annotation.Cacheable
+import org.springframework.stereotype.Component
+
+const val CACHE_NAME_URL_ENTRY = "url-entry-cache"
+
+@Component
+class UrlEntryRepositoryCacheImpl : UrlEntryRepositoryBase {
```

Comment:
Architectural q

1. Why do we cache spring entities? afaiu caches should hold \*actually useful\* data, and an `Entity` is anything but that. It is an oversized monstrosity (contains so much JPA/Spring AOP bloat it scares me) from the POV of a cache.
2. (i never worked with redis + spring, so bear with me) Why is the cache another `Repository`? Is this how Spring best practices suggests we implement caching? Is this how caching is supposed to work? I would like to hear your thoughts here.
3. If this is a `Repository`, why is it annotated with `@Component`?
4. Are we supposed to never use `UrlEntryRepository` again, since we will potentially be missing on caching?

Reply:
1. Not sure what do you mean by that. Every entry in the cache has a key(I'm using the `encoded sequence` for that) and a value(which is a an `UrlEntry`, serialized). On cache hit, that value is DE deserialized back to an UrlEntry. So what data is not useful ?
2. I also never worked before with redis, and definitely didn't handled the caching mechanism in a declarative way, through annotations.  I don't think this class is 'the cache' . Is the decorator which holds and uses the repo, when needed, and adds the caching abstraction on it. Couldn't find an official spring documentation for caching best practices.
3. If you refer to the word `Repository`, the first def on google: `**a place** where or receptacle in which things **are or may be stored**`. If you mean `@Repository`: I don't think is even needed. From spring documentation, about `@Repository`: ` this annotation also serves as a specialization of @Component, **allowing for implementation classes to be autodetected through classpath scanning**` - not the case here. 
4. We are supposed to use both cached/not cached versions

Answer to the last remark: maybe so, but then the UrlService needs to decide by it's own when to cache or not.